lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -298,67 +282,42 @@ def run_async(command)
 
       @started = false
       @user_method.reset
+      @command_pid, @command_in, @command_out = session(get_args(command, with_pty: true), err_stream: false)
+      @started = true
 
-      session.open_channel do |channel|
-        channel.request_pty
-        channel.on_data do |ch, data|
-          publish_data(data, 'stdout') unless @user_method.filter_password?(data)
-          @user_method.on_data(data, ch)
-        end
-        channel.on_extended_data { |ch, type, data| publish_data(data, 'stderr') }
-        # standard exit of the command
-        channel.on_request('exit-status') { |ch, data| publish_exit_status(data.read_long) }
-        # on signal: sending the signal value (such as 'TERM')
-        channel.on_request('exit-signal') do |ch, data|
-          publish_exit_status(data.read_string)
-          ch.close
-          # wait for the channel to finish so that we know at the end
-          # that the session is inactive
-          ch.wait
-        end
-        channel.exec(command) do |_, success|
-          @started = true
-          raise('Error initializing command') unless success
-        end
-      end
-      session.process(0) { !run_started? }
       return true
     end
 
     def run_started?
       @started && @user_method.sent_all_data?
     end
 
+    def read_output_debug(err_io, out_io = nil)
+      stdout = ''
+      debug_str = ''
+
+      if out_io
+        stdout += out_io.read until out_io.eof? rescue
+        out_io.close
+      end
+      debug_str += err_io.read until err_io.eof? rescue
+      err_io.close
+      debug_str.lines.each { |line| @logger.debug(line.strip) }
+
+      return stdout
+    end
+
     def run_sync(command, stdin = nil)
       stdout = ''
       stderr = ''
       exit_status = nil
-      started = false
-
-      channel = session.open_channel do |ch|
-        ch.on_data do |c, data|
-          stdout.concat(data)
-        end
-        ch.on_extended_data { |_, _, data| stderr.concat(data) }
-        ch.on_request('exit-status') { |_, data| exit_status = data.read_long }
-        # Send data to stdin if we have some
-        ch.send_data(stdin) unless stdin.nil?
-        # on signal: sending the signal value (such as 'TERM')
-        ch.on_request('exit-signal') do |_, data|
-          exit_status = data.read_string
-          ch.close
-          ch.wait
-        end
-        ch.exec command do |_, success|
-          raise 'could not execute command' unless success
-
-          started = true
-        end
-      end
-      session.process(0) { !started }
-      # Closing the channel without sending any data gives us SIGPIPE
-      channel.close unless stdin.nil?
-      channel.wait
+
+      pid, tx, rx, err = session(get_args(command))
+      tx.puts(stdin) unless stdin.nil?
+      tx.close
+      stdout = read_output_debug(err, rx)
```

Comment:
capture stdout and stderr returned by read_output_debug, otherwise stderr at L321 will always be an empty string which would make debugging unnecessarily difficult

also you don't need to define stdout, stderr and exit_status on lines 311-313

Reply:
Yeah, I forgot to remove those definitions. Resolved :+1: